### ISSUE TITLE
Smooth Video Loop

### ISSUE THREAD

**Issue by mbzworld** [RenderHeads/UnityPlugin-AVProVideo#400]
When I try to loop a video (4096*4096 30fps H.265 MP4 bitrate 60mbps) the video stutters for like half a second at the beginning of each loop. (Oculus Quest)

Is it possible to avoid this by
- minimizing the video buffer (which might not be possible because if the codec I am using)
 or
- loading the whole video into RAM first? My video size is about 150MB.

**Comment by AndrewRH** [comment]
Hi @mbzworld 

Looping is essentially seeking back to t=0

We have a section in the PDF documentation that explains how to get the best seeking performance.  Basically when it seeks back to 0 it may have to decode several dependent frames before it can start playback.   If you can reduce the complexity of your encoding then this time will be reduced.  The ideal is to have a video with only key-frames.  This will make the file size larger but will give you a smooth loop.  You can use a tool like FFMPEG to do this.  It also has a fast-decode preset that can help.   Basically the video compressors are optimising for playback, and not for fast seeking or looping, so you have to instead encode in a different way to achieve better support for these other use cases.

You could also perhaps try some advanced encoding where you specify all key-frames especially around the end and beginning of the video.

I hope that helps?

Thanks,

**Comment by RenderHeadsSte** [comment]
Small additional that might help your use-case...

What API path are you using?  From experience, it appears seeking is quicker when using MediaPlayer rather than ExoPlayer.

**Comment by mbzworld** [comment]
Thanks Andrew for the suggestion. The problem with having all key-frames video is the file size. It becomes significantly larger or the quality becomes really bad. I tried fast-decode presets before but the stutter was still there. I will experiment more and see if any preset would help.

For the record, I am using the MediaPlayer API path since it was quicker as you said.

Is it possible to load the video file into RAM then play it from there? Do you support such a function?

**Comment by mbzworld** [comment]
Update:
Setting SetPlayWithoutBuffering(false) helped a bit. The freeze is now about 100ms but I noticed some frame drops here and there, the freeze before was something like 300-400ms.

I also tried to use OpenVideoFromBuffer feature after loading the video file into memory:

        byte[] file = System.IO.File.ReadAllBytes(Application.streamingAssetsPath + "/MyVideo.mp4");

        mymediaplayer.OpenVideoFromBuffer(file, true);

        mymediaplayer.Play();

Which I think resulted in even better performance. Probably 50-100ms. 

I thought this might help others as well.
